### ISSUE TITLE
Apply hints when only laziness differs

### ISSUE THREAD

**Issue by josephcsible** [ndmitchell/hlint#723]
If a hint would match a snippet of code except for the presence/absence of `!` or `~`, then we should match the hint anyway, but add `IncreasesLaziness` or `DecreasesLaziness` to the message as appropriate.

**Comment by ndmitchell** [comment]
I guess the only places this triggers are bindings on lambdas?

**Comment by josephcsible** [comment]
I was thinking more so that we could rewrite our rules to be more precise, without making any matches disappear, and without having to use `IncreasesLaziness` or `DecreasesLaziness` manually. For example, I'd like to replace these hints:
```yaml
    - hint: {lhs: "\\(x,y) -> f x y", rhs: uncurry f, note: IncreasesLaziness}
    - hint: {lhs: "\\(x,y) -> (f x, g y)", rhs: Data.Bifunctor.bimap f g, note: IncreasesLaziness}
    - hint: {lhs: "\\(x,y) -> (f x,y)", rhs: Data.Bifunctor.first f, note: IncreasesLaziness}
    - hint: {lhs: "\\(x,y) -> (x,f y)", rhs: Data.Bifunctor.second f, note: IncreasesLaziness}
```
with these ones:
```yaml
    - hint: {lhs: "\\ ~(x,y) -> f x y", rhs: uncurry f}
    - hint: {lhs: "\\ ~(x,y) -> (f x, g y)", rhs: Data.Bifunctor.bimap f g}
    - hint: {lhs: "\\ ~(x,y) -> (f x,y)", rhs: Data.Bifunctor.first f}
    - hint: {lhs: "\\ ~(x,y) -> (x,f y)", rhs: Data.Bifunctor.second f}
```
but still have them continue to match things that they do today.

**Comment by ndmitchell** [comment]
So the advantage would be you get to avoid writing the `note`? And you have slightly more matching power? I worry that if people ever genuinely want to match on lambdas with `~` it's going to be harder.

**Comment by josephcsible** [comment]
I really only want this for the increased matching power. Avoiding writing the `note` is just a bonus.

**Comment by ndmitchell** [comment]
Have you seen lots of places where people actually do this? My guess was that people almost never use `~` annotations on lambda, and thus you wouldn't gain much matching power. But perhaps I'm wrong.
